common/xdc/historyRereplicator.go

```diff
@@ -84,17 +86,27 @@ type (
 		endingNextEventID     int64
 		rereplicator          *HistoryRereplicatorImpl
 		logger                log.Logger
+		ctx                   context.Context
 	}
 )
 
 func newHistoryRereplicationContext(domainID string, workflowID string,
 	beginningRunID string, beginningFirstEventID int64,
 	endingRunID string, endingNextEventID int64,
 	rereplicator *HistoryRereplicatorImpl) *historyRereplicationContext {
+
 	logger := rereplicator.logger.WithTags(
 		tag.WorkflowDomainID(domainID), tag.WorkflowID(workflowID), tag.WorkflowBeginningRunID(beginningRunID),
 		tag.WorkflowBeginningFirstEventID(beginningFirstEventID), tag.WorkflowEndingRunID(endingRunID),
 		tag.WorkflowEndingNextEventID(endingNextEventID))
+
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	timeout := rereplicator.rereplicationTimeout(domainID)
+	if timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
```

Comment:
This doesn't look right. It will be cancelled right after new.

Reply:
Good catch. So I put the cancel into this rereplicaitonContext and have the same lifecycle as the rereplicaitonContext.